drivers/painter/sh1107/qp_sh1107.c

```diff
@@ -0,0 +1,213 @@
+#include "qp_internal.h"
+#include "qp_comms.h"
+#include "qp_surface_internal.h"
+#include "qp_oled_panel.h"
+#include "qp_sh1107.h"
+#include "qp_sh1107_opcodes.h"
+#include "qp_surface.h"
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Driver storage
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+typedef struct sh1107_device_t {
+    oled_panel_painter_device_t oled;
+
+    uint8_t framebuffer[SURFACE_REQUIRED_BUFFER_BYTE_SIZE(128, 128, 1)];
+} sh1107_device_t;
+
+static sh1107_device_t sh1107_drivers[SH1107_NUM_DEVICES] = {0};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Quantum Painter API implementations
+
+// Initialisation
+__attribute__((weak)) bool qp_sh1107_init(painter_device_t device, painter_rotation_t rotation) {
+    sh1107_device_t *driver = (sh1107_device_t *)device;
+
+    // Change the surface geometry based on the panel rotation
+    if (rotation == QP_ROTATION_90 || rotation == QP_ROTATION_270) {
+        driver->oled.surface.base.panel_width  = driver->oled.base.panel_height;
+        driver->oled.surface.base.panel_height = driver->oled.base.panel_width;
+    } else {
+        driver->oled.surface.base.panel_width  = driver->oled.base.panel_width;
+        driver->oled.surface.base.panel_height = driver->oled.base.panel_height;
+    }
+
+    // Init the internal surface
+    if (!qp_init(&driver->oled.surface.base, QP_ROTATION_0)) {
+        qp_dprintf("Failed to init internal surface in qp_sh1107_init\n");
+        return false;
+    }
+
+    // clang-format off
+    uint8_t sh1107_init_sequence[] = {
+        // Command,                 Delay,  N, Data[N]
+        SH1107_SET_MUX_RATIO,           0,  1, 0x7F,      // 1/128 duty
+        SH1107_DISPLAY_OFFSET,          0,  1, 0x00,
+        SH1107_SET_START_LINE,          0,  1, 0x00,      // Different from SH1106
+        SH1107_SET_SEGMENT_REMAP_INV,   0,  0,
+        SH1107_COM_SCAN_DIR_DEC,        0,  0,
+        SH1107_COM_PADS_HW_CFG,         0,  1, 0x12,
+        SH1107_SET_CONTRAST,            0,  1, 0x7F,
+        SH1107_ALL_ON_RESUME,           0,  0,
+        SH1107_NON_INVERTING_DISPLAY,   0,  0,
+        SH1107_SET_OSC_DIVFREQ,         0,  1, 0x80,
+        SH1107_SET_CHARGE_PUMP,         0,  1, 0x14,
+        SH1107_DISPLAY_ON,              0,  0,
+    };
+    // clang-format on
+
+    // If the display height is anything other than the default 128 pixels, change SH1107_SET_MUX_RATIO data byte to the correct value
+    if (driver->oled.base.panel_height != 128) {
+        sh1107_init_sequence[3] = driver->oled.base.panel_height - 1;
+    }
```

Comment:
```suggestion
    // If the display width is anything other than the default 128 pixels, change SH1107_SET_MUX_RATIO data byte to the correct value.
    if (driver->oled.base.panel_width != 128) {
        sh1107_init_sequence[3] = driver->oled.base.panel_width - 1;
    }

    // If the display width is less than the default 128 pixels, change SH1107_DISPLAY_OFFSET to use the center columns.
    if (driver->oled.base.panel_width < 128) {
        sh1107_init_sequence[7] = (128U - driver->oled.base.panel_width) / 2;
    }
```
This change makes the driver work with the 0.96" 64×128 I2C OLED supposedly based on SH1107 (looks  like https://www.aliexpress.us/item/1005004190135420.html, but the one I have had been purchased in 2020, and a new one and the 1.3" version https://www.aliexpress.us/item/1005005473564682.html are still in transit).  It also matches the datasheet (unlike SH1106, in SH1107 the COM pins are used as columns, therefore the mux ratio needs to be changed based on the panel width, not height; and column centering is used for most known displays which have width less than the maximum supported by the controller chip).